static/js/passed-state.js

```diff
@@ -0,0 +1,115 @@
+export default class PassedState {
+    _key = 'python-type-challenges';
+
+    /**
+     * Initializing when there is no state in the local storage. If there is no state in the local storage, the initial state is required.
+     * this function will check the new state and the old state whether is undefined or not and updated the old state based on new state.
+     * 
+     * @param {object} newState - the initial state of the challenges which grouped by the level.
+     * @returns void
+     */
+    init(newState) {
+        const oldState = this.get();
+        // initialize the state when there is no state in the local storage.
+        if (!oldState && !newState) {
+            throw new Error('initial state is required when there is no state in the local storage.');
+        }
+
+        // check new state and old state whether is undefined or not. and merge the new state to the old state.
+        const state = this._checkAndMerge(oldState, newState);
+        this._save(state);
+    }
+
+    get() {
+        const currentState = localStorage.getItem(this._key);
+        return JSON.parse(currentState);
+    }
+
+    /**
+     * Save the state to the local storage with JSON format.
+     * @param {object} state - the state contains the challenge name and whether the challenge is passed.
+     */
+    _save(state) {
+        localStorage.setItem(this._key, JSON.stringify(state));
+    }
+
+    /**
+     * Set the target challenge as passed in the state.
+     * 
+     * @param {'basic' | 'intermediate' | 'advanced' | 'extreme'} level - the level of the challenge.
+     * @param {string} challengeName - the name of the challenge.
+     * @returns void
+     */
+    setPassed(level, challengeName) {
+        let state = this.get();
+
+        const challenges = state[level];
+        for (const challenge of challenges) {
+            if (challenge.name === challengeName) {
+                challenge.passed = true;
+                break;
+            }
+        }
+
+        this._save(state);
+    }
+
+    /**
+     * Merge the new state and the current state. 
+     * this function will compare the new state with the current state and finally overwrite the current state based on the new state: 
+     * - If the old key in the current state isn't in the new state, the old key will be removed from the current state. 
+     * - If the new key in the new state isn't in the current state, the new key will be added to the current state.
+     * 
+     * @param {object} oldState - the current state stored in the local storage.
+     * @param {object} newState - the latest state from the server.
+     * @returns mergedState - the merged state.
+     */
+    _checkAndMerge(oldState, newState) {
```

Comment:
If we use a Map to store the state of challenges, I imagine this method can be simplified

Reply:
If we use Map to store the state of challenges, we still to convert it to object when serialization and storing to local storage.

`JSON.stringify` doesn't support to serialize Map directly.

Reply:
Yep, looks like conversion is needed
https://stackoverflow.com/a/28918362/2142577

Reply:
So, should I leave state as is? 🤔

Reply:
SGTM